Improve security docs (#1103)
solidjs/solid-docs: src/routes/solid-start/guides/security.mdx

title: Security
---

## XSS (Cross Site Scripting)

Solid automatically escape values passed to JSX expressions to reduce the risk of XSS attacks.

> However, this protection does not apply when using [`innerHTML`](https://docs.solidjs.com/reference/jsx-attributes/innerhtml-or-textcontent#innerhtml-or-textcontent).

Reviewer: ```suggestion
However, this protection does not apply when using [`innerHTML`](/reference/jsx-attributes/innerhtml-or-textcontent#innerhtml-or-textcontent).
```

title: Security
---

> This guide walks you through how to implement common security mechanisms in SolidStart.

Reviewer: ```suggestion
```

## XSS (Cross Site Scripting)

Solid automatically escape values passed to JSX expressions to reduce the risk of XSS attacks.
However, this protection does not apply when using [`innerHTML`](https://docs.solidjs.com/reference/jsx-attributes/innerhtml-or-textcontent#innerhtml-or-textcontent).

To protect your application from XSS attacks:

> - Avoid using [`innerHTML`](/reference/jsx-attributes/innerhtml-or-textcontent#innerhtml-or-textcontent) when possible.

Reviewer: redundant link
```suggestion
- Avoid using `innerHTML` when possible.
```

Author: why redundant?

Reviewer: It's mentioned two lines above

  If necessary, make sure to sanitize user-supplied data with libraries such as [DOMPurify](https://github.com/cure53/DOMPurify).

> - Validate and sanitize user inputs.
>   Always validate form inputs on the server in addition to the client.

Reviewer: ```suggestion
- Validate and sanitize user inputs, especially form inputs on the server and client.
```

> - Set a Content Security Policy (CSP).

Reviewer: could you link this to where it is in the mdn

Author: added a link to CSP docs in MDN

- Sanitize attributes containing user-supplied data within `<noscript>` elements.
  This includes both the attributes of the `<noscript>` element itself and its children.

> We highly recommend reading the [Cross Site Scripting Prevention Cheat Sheet](https://cheatsheetseries.owasp.org/cheatsheets/Cross_Site_Scripting_Prevention_Cheat_Sheet.html) for further guidance.

Reviewer: ```suggestion
It is highly recommend to reading the [Cross Site Scripting Prevention Cheat Sheet](https://cheatsheetseries.owasp.org/cheatsheets/Cross_Site_Scripting_Prevention_Cheat_Sheet.html) for further guidance.
```

> ## CSP (Content Security Policy)

Reviewer: ```suggestion
## Content Security Policy (CSP)
```

> To configure the `Content-Security-Policy` HTTP header, you can use a [middleware](/solid-start/advanced/middleware).

Reviewer: ```suggestion
To configure the `Content-Security-Policy` HTTP header, [middleware](/solid-start/advanced/middleware) can be used.
```

### With nonce (recommended)

> If you want to use a strict CSP with nonces:

Reviewer: Can you add a bit of a description to nonces plus a link to the MDN where it talks more about it

Author: Added a link to MDN docs about strict CSP and added a step about generating the nonce.

> 1. Create a middleware that configures the CSP header, then register it to run on the [`onRequest`](/solid-start/advanced/middleware#onrequest) event.

Reviewer: ```suggestion
1. Create a middleware that configures the CSP header.
It can then be registered using the [`onRequest`](/solid-start/advanced/middleware#onrequest) event.
```

Author: I would like to keep a more direct tone. In this case, the middleware **must** be registered in `onRequest`,  otherwise the nonce won't be available in the `entry-server.tsx` file. Saying "it can be registered" implies that there are multiple options, but that is not the case.

Reviewer: Matter of changing a word - done.

```suggestion
1. Create a middleware that configures the CSP header.
It must then be registered using the [`onRequest`](/solid-start/advanced/middleware#onrequest) event.
```

2. Store the nonce in the [`locals`](/solid-start/advanced/middleware#locals) object.
3. Configure SolidStart to use the nonce in your [`entry-server.tsx`](/solid-start/reference/entrypoints/entry-server) file.

<TabsCodeBlocks>
<div id="Middleware">

```tsx
import { createMiddleware } from "@solidjs/start/middleware";
import { randomBytes } from "crypto";

export default createMiddleware({
	onRequest: (event) => {
		const nonce = randomBytes(16).toString("base64");

		event.locals.nonce = nonce;

		const csp = `
      default-src 'self';
      script-src 'nonce-${nonce}' 'strict-dynamic' 'unsafe-eval';
      object-src 'none';
      base-uri 'none';
      frame-ancestors 'none';
      form-action 'self';
    `.replace(/\s+/g, " ");

		event.response.headers.set("Content-Security-Policy", csp);
	},
});
```

</div>
<div id="entry-server.tsx">

```tsx {6} title="src/entry-server.tsx"
// @refresh reload
import { createHandler, StartServer } from "@solidjs/start/server";

export default createHandler(
	() => <StartServer /* ... */ />,
	(event) => ({ nonce: event.locals.nonce })
);
```

</div>
</TabsCodeBlocks>

### Without nonce

> To configure CSP without a nonce, create a middleware that configures the CSP header, then register it to run on the [`onBeforeResponse`](/solid-start/advanced/middleware#onbeforeresponse) event:

Reviewer: ```suggestion
To configure CSP without a nonce, a middleware that sets the CSP header is required, and it should be registered to run during the [`onBeforeResponse`](/solid-start/advanced/middleware#onbeforeresponse) event:
```

```tsx
import { createMiddleware } from "@solidjs/start/middleware";

export default createMiddleware({
	onBeforeResponse: (event) => {
		const csp = `
      default-src 'self';
      font-src 'self'  ;
      object-src 'none';
      base-uri 'none';
      frame-ancestors 'none';
      form-action 'self';
    `.replace(/\s+/g, " ");

		event.response.headers.set("Content-Security-Policy", csp);
	},
});
```

## CORS (Cross-Origin Resource Sharing)

> When you want to allow other applications to access your API endpoints, you can use a middleware to configure the CORS headers:

Reviewer: ```suggestion
When other applications need access to API endpoints, a middleware that configures the CORS headers is needed:
```

import { json } from "@solidjs/router";

const TRUSTED_ORIGINS = ["https://my-app.com", "https://another-app.com"];

export default createMiddleware({
	onBeforeResponse: (event) => {
		const { request, response } = event;

		response.headers.append("Vary", "Origin, Access-Control-Request-Method");

		const origin = request.headers.get("Origin");
		const requestUrl = new URL(request.url);
		const isApiRequest = requestUrl && requestUrl.pathname.startsWith("/api");

		if (isApiRequest && origin && TRUSTED_ORIGINS.includes(origin)) {
			// Handle preflight requests.
			if (
				request.method === "OPTIONS" &&
				request.headers.get("Access-Control-Request-Method")
			) {
				// Preflight requests are standalone, so we immediately send a response.
				return json(null, {
					headers: {
						"Access-Control-Allow-Origin": origin,
						"Access-Control-Allow-Methods": "OPTIONS, POST, PUT, PATCH, DELETE",
						"Access-Control-Allow-Headers": "Authorization, Content-Type",
					},
				});
			}

			// Handle normal requests.
			response.headers.set("Access-Control-Allow-Origin", origin);
		}
	},
});
```

## CSRF (Cross-Site Request Forgery)

> To prevent CSRF attacks, you can use a middleware to block untrusted requests:

Reviewer: ```suggestion
To prevent CSRF attacks, a middleware can be used to block untrusted requests:
```

			const requestUrl = new URL(request.url);
			const origin = request.headers.get("Origin");

			// If we have an Origin header, check it against our allowlist.
			if (origin) {
				const parsedOrigin = new URL(origin);

				if (
					parsedOrigin.origin !== requestUrl.origin &&
					!TRUSTED_ORIGINS.includes(parsedOrigin.host)
				) {
					return json({ error: "origin invalid" }, { status: 403 });
				}
			}

			// If we are serving via TLS and have no Origin header, prevent against
			// CSRF via HTTP man-in-the-middle attacks by enforcing strict Referer
			// origin checks.
			if (!origin && requestUrl.protocol === "https:") {
				const referer = request.headers.get("Referer");

				if (!referer) {
					return json({ error: "referer not supplied" }, { status: 403 });
				}

				const parsedReferer = new URL(referer);

				if (parsedReferer.protocol !== "https:") {
					return json({ error: "referer invalid" }, { status: 403 });
				}

				if (
					parsedReferer.host !== requestUrl.host &&
					!TRUSTED_ORIGINS.includes(parsedReferer.host)
				) {
					return json({ error: "referer invalid" }, { status: 403 });
				}
			}
		}
	},
});
```

> This example demonstrates a basic implementation of CSRF protection that checks for the `Origin` and `Referer` headers and blocks requests that are not from trusted origins.

Reviewer: ```suggestion
This example demonstrates a basic CSRF protection that verifies the `Origin` and `Referer` headers, blocking requests from untrusted origins.
```

> In addition, consider implementing a more comprehensive CSRF protection mechanism such as [Double-Submit Cookie Pattern](https://cheatsheetseries.owasp.org/cheatsheets/Cross-Site_Request_Forgery_Prevention_Cheat_Sheet.html#alternative-using-a-double-submit-cookie-pattern).

Reviewer: ```suggestion
Additionally, consider implementing a more robust CSRF protection mechanism, such as the [Double-Submit Cookie Pattern](https://cheatsheetseries.owasp.org/cheatsheets/Cross-Site_Request_Forgery_Prevention_Cheat_Sheet.html#alternative-using-a-double-submit-cookie-pattern).
```

> We highly recommend reading the [Cross-Site Request Forgery Prevention Cheat Sheet](https://cheatsheetseries.owasp.org/cheatsheets/Cross-Site_Request_Forgery_Prevention_Cheat_Sheet.html) for further guidance.

Reviewer: ```suggestion
For further guidance, you can look to the [Cross-Site Request Forgery Prevention Cheat Sheet](https://cheatsheetseries.owasp.org/cheatsheets/Cross-Site_Request_Forgery_Prevention_Cheat_Sheet.html).
```
